### PR TITLE
Adding data source and resource changes for dns forward zone

### PR DESCRIPTION
**Pull request by Arvinkumar3**
* Added data source and resource changes for dns forward zone
* Changes are tested in env-2a and it is working fine.

# Issue/Feature description

* 

# How it was fixed/implemented

* 

# Tests

* 

# Reviewers

### CODE REVIEW

**Comment by venkat-iblox** [comment]
Please don't use internal references on PRs to this public repo. You can leave the title as `adding DNS Forward zone support`

**Comment by akleinib** [comment]
Do you plan to add documentation under `docs/resources` and `docs/data-sources` in a subsequent PR?

File: b1ddi/resource_dns_forward_zone.go [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+			},
+		},
+	}
+}
+
+func resourceConfigForwardZoneCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	c := m.(*b1ddiclient.Client)
+
+	externalForwarders := make([]*models.ConfigForwarder, 0)
+	for _, ep := range d.Get("external_forwarders").([]interface{}) {
```
Comment (akleinib) [inline b1ddi/resource_dns_forward_zone.go:180]: I guess `ep` comes from a copy/paste from auth zone and external primaries. Using `ef` would be better.

File: b1ddi/resource_dns_forward_zone.go [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+	err = d.Set("created_at", resp.Payload.Result.CreatedAt.String())
+	if err != nil {
+		diags = append(diags, diag.FromErr(err)...)
+	}
+	err = d.Set("disabled", resp.Payload.Result.Disabled)
+	if err != nil {
+		diags = append(diags, diag.FromErr(err)...)
+	}
+	externalForwarders := make([]map[string]interface{}, 0, len(resp.Payload.Result.ExternalForwarders))
+	for _, ep := range resp.Payload.Result.ExternalForwarders {
```
Comment (akleinib) [inline b1ddi/resource_dns_forward_zone.go:264]: As above.

File: b1ddi/resource_dns_forward_zone.go [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+		return diag.FromErr(fmt.Errorf("changing the value of 'fqdn' field is not allowed"))
+	}
+
+	if d.HasChange("view") {
+		d.Partial(true)
+		return diag.FromErr(fmt.Errorf("changing the value of 'view' field is not allowed"))
+	}
+
+	externalForwarders := make([]*models.ConfigForwarder, 0)
+	for _, ep := range d.Get("external_forwarders").([]interface{}) {
```
Comment (akleinib) [inline b1ddi/resource_dns_forward_zone.go:332]: As above.

File: b1ddi/resource_dns_forward_zone.go [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+
+	if d.HasChange("fqdn") {
+		d.Partial(true)
+		return diag.FromErr(fmt.Errorf("changing the value of 'fqdn' field is not allowed"))
+	}
+
+	if d.HasChange("view") {
+		d.Partial(true)
+		return diag.FromErr(fmt.Errorf("changing the value of 'view' field is not allowed"))
+	}
```
Comment (akleinib) [inline b1ddi/resource_dns_forward_zone.go:329]: Why only checking these two fields and not the remaining ones?

File: b1ddi/data_source_dns_forward_zone.go [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+	resp, err := c.DNSConfigurationAPI.ForwardZone.ForwardZoneList(&forward_zone.ForwardZoneListParams{
+		Filter:  swag.String(filterStr),
+		Context: ctx,
+	}, nil)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	results := make([]interface{}, 0, len(resp.Payload.Results))
+	for _, ab := range resp.Payload.Results {
```
Comment (akleinib) [inline b1ddi/data_source_dns_forward_zone.go:53]: Do we use `ab` for Address Block?

File: b1ddi/data_source_dns_forward_zone.go [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+	return diags
+}
+
+func flattenConfigForwardZone(r *models.ConfigForwardZone) []interface{} {
+	if r == nil {
+		return nil
+	}
+
+	externalForwarders := make([]map[string]interface{}, 0, len(r.ExternalForwarders))
+	for _, ep := range r.ExternalForwarders {
```
Comment (akleinib) [inline b1ddi/data_source_dns_forward_zone.go:73]: I guess `ep` comes from a copy/paste from auth zone and external primaries. Using `ef` would be better.

File: examples/data-sources/b1ddi_dns_forward_zones/data-source.tf [hunk 7]
```diff
@@ -0,0 +1,18 @@
+terraform {
+  required_providers {
+    b1ddi = {
+      version = "0.1"
```
[the remark this answers is not in the document]
Comment (Arvinkumar3) [reply examples/data-sources/b1ddi_dns_forward_zones/data-source.tf]: Removed version. Terraform will take the latest version. However as discussed, version has to be '0.1.5' as '0.1.4' does not support forward zone.

File: examples/resources/b1ddi_dns_forward_zone/resource.tf [hunk 8]
```diff
@@ -0,0 +1,17 @@
+terraform {
+  required_providers {
+    b1ddi = {
+      version = "0.1"
```
[the remark this answers is not in the document]
Comment (Arvinkumar3) [reply examples/resources/b1ddi_dns_forward_zone/resource.tf]: Removed version.

File: examples/resources/b1ddi_dns_forward_zone/resource.tf [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+  }
+}
+
+resource "b1ddi_dns_view" "tf_example_dns_view" {
+  name = "example_tf_dns_view"
+}
+
+resource "b1ddi_dns_forward_zone" "tf_example_forward_zone" {
+  fqdn = "tf-example.com."
+  view = b1ddi_dns_view.tf_example_dns_view.id
```
[the remark this answers is not in the document]
Comment (Arvinkumar3) [reply examples/resources/b1ddi_dns_forward_zone/resource.tf:28]: Added host and tested the changes in env-2a. A functional forward zone got created after adding host info.

File: b1ddi/resource_dns_forward_zone.go (hunk 1 above)
Comment (Arvinkumar3) [reply b1ddi/resource_dns_forward_zone.go]: Updated

File: b1ddi/resource_dns_forward_zone.go (hunk 3 above)
Comment (Arvinkumar3) [reply b1ddi/resource_dns_forward_zone.go]: Updated

File: b1ddi/resource_dns_forward_zone.go [hunk 10]
```diff
@@ -0,0 +1,395 @@
...
+
+	body := &models.ConfigForwardZone{
+		Comment:            d.Get("comment").(string),
+		Disabled:           d.Get("disabled").(bool),
+		ExternalForwarders: externalForwarders,
+		ForwardOnly:        d.Get("forward_only").(bool),
+		Hosts:              hosts,
+		InternalForwarders: internalForwarders,
+		Nsgs:               nsgs,
+		Parent:             d.Get("parent").(string),
```
[the remark this answers is not in the document]
Comment (Arvinkumar3) [reply b1ddi/resource_dns_forward_zone.go]: Removed.

File: b1ddi/resource_dns_forward_zone.go (hunk 4 above)
Comment (Arvinkumar3) [reply b1ddi/resource_dns_forward_zone.go:328]: As only these two fields cannot be altered for a forward zone.

File: b1ddi/data_source_dns_forward_zone.go (hunk 6 above)
Comment (Arvinkumar3) [reply b1ddi/data_source_dns_forward_zone.go]: Updated

File: b1ddi/data_source_dns_forward_zone.go (hunk 5 above)
Comment (Arvinkumar3) [reply b1ddi/data_source_dns_forward_zone.go]: Updated

**Comment by Arvinkumar3** [comment]
> Do you plan to add documentation under `docs/resources` and `docs/data-sources` in a subsequent PR?

@akleinib Yes, It will be taken care in a separate PR.

**Review by akleinib — APPROVED** [review]

**Pull request merged**